### PR TITLE
Merging Packet changes

### PR DESCRIPTION
**Pull request by Kikit24**
Added Packet class with Packet buffer serialization and read/write support for byte, int, float, double, and GLM vec3.

### CODE REVIEW

**Comment by tlacaba** [comment]
As far as I know looks good, but I might need to get a tutorial on how to use it!

File: common/packet/include/packet.hpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#include <winsock2.h>
+#else
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netinet/in.h>
+#include <unistd.h>
+#include <string.h>
+#include <vector>
+#endif
```
Comment (shanekim28) [inline common/packet/include/packet.hpp]: Probably don't need all this

File: common/packet/include/packet.hpp [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+class Packet {
+
+   public:
+    std::vector<u_int8_t> buffer;
```
Comment (shanekim28) [inline common/packet/include/packet.hpp]: use `uint8_t` instead of `u_int8_t` for all occurrences

File: common/packet/main.cpp
Comment (shanekim28) [inline common/packet/main.cpp:1]: This file probably shouldn't be added in this merge, or it should be put in a separate tests directory

File: common/packet/src/packet.cpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    u_int64_t l;
+} union64;
+
+Packet::Packet() {
+    this->buffer = std::vector<u_int8_t>();
+}
+
+Packet::~Packet() {
+    this->buffer = std::vector<u_int8_t>();
+}
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]:
```suggestion
Packet::Packet() {
    this->buffer = std::vector<u_int8_t>();
}

Packet::~Packet() {
    this->buffer = std::vector<u_int8_t>();
}
```
Unneeded, the default constructor/destructor already handles this. If you're gonna add a custom destructor, see [rule of three](https://en.wikipedia.org/wiki/Rule_of_three_(C%2B%2B_programming)#:~:text=The%20rule%20of%20three%20(also,copy%20assignment%20operator)

File: common/packet/include/packet.hpp (hunk 2 above)
Comment (shanekim28) [inline common/packet/include/packet.hpp]: Also use `deque` instead of `vector`

File: common/packet/src/packet.cpp [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    write_float(data.z);
+}
+
+int Packet::read_byte(char* dest){
+    if(this->buffer.size() == 0){
+        return -1;
+    }
+
+    u_int8_t readElem = this->buffer.front();
+    this->buffer.erase(this->buffer.begin());
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]:
```suggestion
buffer.pop_front();
```

File: common/packet/src/packet.cpp [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    *dest = char(readElem);
+    
+    return 1;
+}
+
+int Packet::read_int(int* dest){
+    int int8_per_int = 4;
+
+    if(this->buffer.size() < int8_per_int){
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]: don't need `this`

File: common/packet/src/packet.cpp [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+
+    u_int8_t readElem = this->buffer.front();
+    this->buffer.erase(this->buffer.begin());
+    *dest = char(readElem);
+    
+    return 1;
+}
+
+int Packet::read_int(int* dest){
+    int int8_per_int = 4;
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]: Bad, just use `sizeof(int)`

File: common/packet/src/packet.cpp [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+    if(this->buffer.size() < int8_per_int){
+        return -1;
+    }
+
+    // read 4 bytes from the buffer
+    u_int32_t readData[int8_per_int];
+    for(int i = 0; i < int8_per_int; i++){
+        readData[i] = this->buffer.front();
+        this->buffer.erase(this->buffer.begin());
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]:
```suggestion
buffer.pop_front();
```
Don't need `this` as well

File: common/packet/src/packet.cpp [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+    }
+
+    // reconstruct u_int32
+    u_int32_t i32 = readData[3] | (readData[2] << 8) | (readData[1] << 16) | (readData[0] << 24);
+    i32 = ntohl(i32);
+
+    *dest = static_cast<int>(i32); // convert to int
+
+    return int8_per_int * 8;
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]:
```suggestion
sizeof(int)
```

File: common/packet/src/packet.cpp [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    // reconstruct u_int32
+    u_int32_t i32 = readData[3] | (readData[2] << 8) | (readData[1] << 16) | (readData[0] << 24);
+    i32 = ntohl(i32);
+
+    // convert to float
+    union32.l = i32;
+    *dest = (float)union32.f;
+
+    return int8_per_float * 8;
+}
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]: See comments on `read_int`

File: common/packet/src/packet.cpp [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+    // reconstruct u_int64
+    u_int64_t i64 = readData[7] | (readData[6] << 8) | (readData[5] << 16) | (readData[4] << 24) | (readData[3] << 32) | (readData[2] << 40) | (readData[1] << 48) |  (readData[0] << 56);
+
+    // convert to double
+    i64 = ntoh64(&i64);
+    union64.l = i64;
+    *dest = union64.f;
+
+    return int8_per_double * 8;
+}
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]: See comments on `read_int`

File: common/packet/src/packet.cpp [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+        return -1;
+    }
+    if(read_float(&z) == -1){
+        return -1;
+    }
+
+    // reconstruct vector
+    *dest = glm::vec3(x, y, z);
+
+    return 32*3;
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]:
```suggestion
return sizeof(float) * 3
```

File: common/packet/src/packet.cpp [hunk 12]
```diff
@@ -0,0 +1,239 @@
...
+
+
+
+u_int8_t* Packet::getBytes() {
+    u_int8_t* byte_array = (u_int8_t*)malloc(this->buffer.size() * sizeof(u_int8_t));
+    int i = 0;
+
+    for(u_int8_t b : this->buffer){
+       byte_array[i++] = b;
+    }
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]:
```suggestion
    uint8_t* byte_array = new uint8_t[buffer.size()];
    memcpy(byte_array, &buffer[0], buffer.size());
```

**Review by shanekim28 — CHANGES_REQUESTED** [review]
Good structure, just some changes needed

File: common/packet/src/packet.cpp (hunk 3 above)
Comment (Kikit24) [reply common/packet/src/packet.cpp]: Good point, thanks

File: common/packet/main.cpp
Comment (Kikit24) [reply common/packet/main.cpp:1]: Done

File: tests/packet/Packet
Comment (shanekim28) [inline tests/packet/Packet:1]: We shouldn't be including binary files

File: common/packet/include/packet.hpp [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+    void write_vec3(glm::vec3 data);
+
+    // return -1 if error, else return size
+    int read_byte(char* dest);
+    int read_int(int* dest);
+    int read_float(float* dest);
+    int read_double(double* dest);
+    int read_vec3(glm::vec3* dest);
+
+    u_int8_t* getBytes();
```
Comment (shanekim28) [inline common/packet/include/packet.hpp]: Make sure you change every instance of u_int8_t to uint8_t, u_int32_t to uint32_t, etc.

File: common/packet/src/packet.cpp [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+} union32;
+
+union Union64 {
+    double f;
+    u_int64_t l;
+} union64;
+
+void Packet::write_byte(char data){
+    u_int8_t toAdd = u_int8_t(data);
+    buffer.push_back(toAdd);
```
Comment (shanekim28) [inline common/packet/src/packet.cpp]:
```suggestion
    buffer.push_back((uint8_t) data);
```
The function already makes a copy of the parameter and puts it on the stack, no need to make another copy.

File: common/packet/include/packet.hpp [hunk 15]
```diff
@@ -0,0 +1,30 @@
+#ifndef PACKET_H
+#define PACKET_H
+#include <stdio.h>
+#include <iostream>
```
Comment (shanekim28) [inline common/packet/include/packet.hpp]: Do we really need these in Packet? I don't think anything's being printed

**Review by shanekim28 — DISMISSED** [review]
Great! Just a couple more changes and it's good to go

File: tests/packet/Packet
Comment (Kikit24) [reply tests/packet/Packet:1]: Oops I thought I did git rm on that file. I didn’t realize it committed

File: common/packet/include/packet.hpp (hunk 15 above)
Comment (Kikit24) [reply common/packet/include/packet.hpp]: You’re right - I deleted all the print statements.

**Pull request merged**